utils/constraints/src/lib.rs

```diff
@@ -0,0 +1,245 @@
+        Constraints {
+            constraints: self
+                .into_iter()
+                .filter(|c| match c {
+                    ConstraintExpr::KeyValue { key, .. } => *key != del_key,
+                    _ => true,
+                })
+                .collect(),
+            operator: op,
+        }
+    }
+    pub fn filter_by_key<T: Into<ConstraintKey>>(&self, get_key: T) -> Option<Constraints> {
+        let k = get_key.into();
+        let v: Vec<_> = self
+            .constraints
+            .iter()
+            .cloned()
+            .filter(|e| match e {
+                ConstraintExpr::KeyValue { key, .. } => *key == k,
+                ConstraintExpr::Constraints(_) => false,
+            })
+            .collect();
+        match v.len() {
+            0 => None,
+            1 => Some(Self::new_single(v[0].clone())),
+            _ => Some(Self::new_clause(self.operator, v)),
+        }
+    }
+}
+
+impl fmt::Display for Constraints {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self.constraints.len() {
+            0 => Ok(()),
+            1 => write!(f, "{}", self.constraints[0]),
+            _ => {
+                write!(f, "({}\n", self.operator.to_string())?;
+                for el in &self.constraints {
+                    write!(f, "  {}\n", el.to_string().replace("\n", "\n  "))?;
+                }
+                write!(f, ")")
+            }
+        }
+    }
+}
+
+impl std::iter::IntoIterator for Constraints {
+    type Item = ConstraintExpr;
+    type IntoIter = std::vec::IntoIter<Self::Item>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.constraints.into_iter()
+    }
+}
+
+#[derive(Copy, Clone, PartialEq)]
+pub enum ClauseOperator {
+    And,
+    Or,
+}
+
+impl fmt::Display for ClauseOperator {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{}",
+            match self {
+                ClauseOperator::And => "&",
+                ClauseOperator::Or => "|",
+            }
+        )
+    }
+}
+
+#[derive(Copy, Clone)]
+pub enum ConstraintOperator {
+    Equal,
+    NotEqual,
+    LessThan,
+    GreaterThan,
+}
+
+impl fmt::Display for ConstraintOperator {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{}",
+            match self {
+                ConstraintOperator::Equal => "=",
+                ConstraintOperator::NotEqual => "<>",
+                ConstraintOperator::LessThan => "<",
+                ConstraintOperator::GreaterThan => ">",
+            }
+        )
+    }
+}
+
+#[derive(Clone, PartialEq)]
+pub struct ConstraintKey(serde_json::Value);
+
+impl ConstraintKey {
+    pub fn new<T: Into<serde_json::Value>>(v: T) -> Self {
+        ConstraintKey(v.into())
+    }
+}
+
+impl From<&str> for ConstraintKey {
+    fn from(key: &str) -> Self {
+        ConstraintKey::new(serde_json::Value::String(key.to_string()))
+    }
+}
```

Comment:
Probably you would like to use AsRef https://doc.rust-lang.org/std/convert/trait.AsRef.html to catch everything what is stringable

Reply:
This method is for creating ConstraintKey from key: &str, so it's better if it owns the key instead of having a reference to it, but I'm not sure.

Reply:
Thats not what I meant. If you will write signature:
```
impl From<T: AsRef<str>> for ConstraintKey
...
```
You can write key.as_ref().to_string() and create new object as you do now.
But
advantage is, that you can call ConstraintKey::from passing as parameter whatever implements AsRef for str.
But it's just a suggestion, I don't know if you need it

Reply:
You are right. This was added mostly for &str (as library users would be directly typing &strs), but I'm going to improve this to handle everything that can be treated as a reference to str now.

Reply:
Done.